TOOLS/umpv

```diff
@@ -26,62 +26,78 @@ Note: you can supply custom mpv path and options with the MPV environment
       is already running), this will be ignored.
 """
 
-import sys
 import os
-import socket
-import errno
-import subprocess
-import string
 import shlex
+import string
+import subprocess
+import sys
+from collections.abc import Iterable
 
-files = sys.argv[1:]
 
-# this is the same method mpv uses to decide this
-def is_url(filename):
+def is_url(filename: str) -> bool:
     parts = filename.split("://", 1)
     if len(parts) < 2:
         return False
     # protocol prefix has no special characters => it's an URL
-    allowed_symbols = string.ascii_letters + string.digits + '_'
+    allowed_symbols = string.ascii_letters + string.digits + "_"
     prefix = parts[0]
-    return all(map(lambda c: c in allowed_symbols, prefix))
-
-# make them absolute; also makes them safe against interpretation as options
-def make_abs(filename):
-    if not is_url(filename):
-        return os.path.abspath(filename)
-    return filename
-files = (make_abs(f) for f in files)
-
-SOCK = os.path.join(os.getenv("XDG_RUNTIME_DIR", os.getenv("HOME")), ".umpv_socket")
-
-sock = None
-try:
-    sock = socket.socket(socket.AF_UNIX)
-    sock.connect(SOCK)
-except socket.error as e:
-    if e.errno == errno.ECONNREFUSED:
-        sock = None
-        pass  # abandoned socket
-    elif e.errno == errno.ENOENT:
-        sock = None
-        pass # doesn't exist
-    else:
-        raise e
-
-if sock:
-    # Unhandled race condition: what if mpv is terminating right now?
-    for f in files:
-        # escape: \ \n "
-        f = f.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n")
-        f = "\"" + f + "\""
-        sock.send(("raw loadfile " + f + " append\n").encode("utf-8"))
-else:
-    # Let mpv recreate socket if it doesn't already exist.
-
-    opts = shlex.split(os.getenv("MPV") or "mpv")
-    opts.extend(["--no-terminal", "--force-window", "--input-ipc-server=" + SOCK,
-                 "--"])
-    opts.extend(files)
-
-    subprocess.check_call(opts)
+    return all(c in allowed_symbols for c in prefix)
+
+def get_socket_path() -> str:
+    if os.name == "nt":
+        return r"\\.\pipe\umpv"
+
+    base_dir = (
+        os.getenv("UMPV_SOCKET_DIR") or
+        os.getenv("XDG_RUNTIME_DIR") or
+        os.getenv("HOME") or
+        os.getenv("TMPDIR")
+    )
+
+    if not base_dir:
+        raise Exception("Could not determine a base directory for the socket. "
+                        "Ensure that one of the following environment variables is set: "
+                        "UMPV_SOCKET_DIR, XDG_RUNTIME_DIR, HOME or TMPDIR.")
+
+    return os.path.join(base_dir, ".umpv")
+
+def send_files_to_mpv(conn, files: Iterable[str]) -> None:
+    try:
+        send = conn.send if os.name != "nt" else conn.write
```

Comment:
pythonic solution, probably
```suggestion
        send = hasattr(conn, "send) and conn.send or conn.write
```

Reply:
I've used `isinstance(conn, socket.socket)` to be more explicit, let me know if that's ok.

Reply:
I think that disagrees with the duck typing principle but it also does not matter.